src/osd/PrimaryLogPG.h

```diff
@@ -1623,17 +1623,30 @@ class PrimaryLogPG : public PG, public PGBackend::Listener {
     struct ReservationCB : public Context {
       PrimaryLogPGRef pg;
       bool canceled;
-      ReservationCB(PrimaryLogPG *pg) : pg(pg), canceled(false) {}
+      Mutex lock;
+      ReservationCB(PrimaryLogPG *pg) : pg(pg), canceled(false),
+					lock("ReservationCB", false, false) {}
       void finish(int) override {
-	pg->lock();
-	if (!canceled)
-	  pg->snap_trimmer_machine.process_event(SnapTrimReserved());
-	pg->unlock();
+	PrimaryLogPGRef pgp;
+	bool is_canceled = false;
+	{
+	  Mutex::Locker l(lock);
+	  pgp = pg;
+	  is_canceled = canceled;
+	}
+	if (!pgp)
+	  return;
+	pgp->lock();
+	if (!is_canceled)
+	  pgp->snap_trimmer_machine.process_event(SnapTrimReserved());
+	pgp->unlock();
       }
       void cancel() {
 	assert(pg->is_locked());
+	Mutex::Locker l(lock);
 	assert(!canceled);
 	canceled = true;
+	pg.reset();
```

Comment:
It seems like this doesn't reliably make the ref go away.  The asyncreserver could grab pgp, drop the local lock, then on_shutdown runs under pg lock and cancels.  What ensures that asyncreserver is not still  blocked on pg->lock (or scheduled away) when we assert there are no refs?  Do we flush the asyncserver thread or something?  (And if we do flush the asyncreserver, can't we make sure the Context goes away that way and avoid this change entirely?)